sui_programmability/examples/regulated_coin/sources/RegulatedCoin.move

```diff
@@ -0,0 +1,225 @@
+        to: address
+    }
+
+    /// Currently let's just use Coin::TreasuryCap functionality
+    fun init(ctx: &mut TxContext) {
+        Transfer::transfer(
+            Coin::create_currency(REST {}, ctx),
+            TxContext::sender(ctx)
+        )
+    }
+
+    /// Only owner of the treasury cap can create new Balances; for example, after a KYC process;
+    public fun create_empty_for(_cap: &TreasuryCap<REST>, for: address, ctx: &mut TxContext) {
+        Transfer::transfer(C::zero(REST {}, for, ctx), for)
+    }
+
+    /// Allow borrowing as is, by default
+    public fun borrow(coin: &RegulatedCoin<REST>): &Balance<REST> { C::borrow(REST {}, coin) }
+    public fun borrow_mut(coin: &mut RegulatedCoin<REST>, ctx: &mut TxContext): &mut Balance<REST> {
+        assert!(TxContext::sender(ctx) == C::owner(coin), ENotOwner); // only owner can access the balance
+        C::borrow_mut(REST {}, coin)
+    }
+
+    // === Coin Transfers ===
+
+    public(script) fun transfer(
+        coin: &mut RegulatedCoin<REST>, value: u64, to: address, ctx: &mut TxContext
+    ) {
+        Transfer::transfer(CoinTransfer {
+            id: TxContext::new_id(ctx),
+            balance: Balance::split(borrow_mut(coin, ctx), value),
+            to
+        }, to)
+    }
+
+    public(script) fun accept_transfer(
+        coin: &mut RegulatedCoin<REST>, transfer: CoinTransfer, ctx: &mut TxContext
+    ) {
+        let CoinTransfer { id, balance, to } = transfer;
+        assert!(C::owner(coin) == to, ENotOwner);
+        Balance::join(borrow_mut(coin, ctx), balance);
+        Sui::ID::delete(id);
+    }
+
+    // === Explicit "Not Implemented" part ===
+
+    public fun join() { abort ENotImplemented }
+    public fun split() { abort ENotImplemented }
+    public fun into_balance() { abort ENotImplemented }
+    public fun from_balance() { abort ENotImplemented }
+}
+
+module Stable::RestrictedStake {
+    use Stable::REST::{Self, REST};
+    use Stable::RegulatedCoin::RegulatedCoin;
+
+    use Sui::Coin::{Self, Coin, TreasuryCap};
+    use Sui::Balance::{Self, Balance};
+    use Sui::TxContext::{Self, TxContext};
+    use Sui::Transfer;
+
+    // stake token - get your money back once
+    struct STAKE has drop {}
+
+    struct StableStake has key {
+        id: Sui::ID::VersionedID,
+        balance: Balance<REST>,
+        treasury_cap: TreasuryCap<STAKE>,
+    }
+
+    fun init(ctx: &mut TxContext) {
+        Transfer::share_object(StableStake {
+            id: TxContext::new_id(ctx),
+            balance: Balance::zero<REST>(),
+            treasury_cap: Coin::create_currency<STAKE>(STAKE{}, ctx)
+        });
+    }
+
+    public(script) fun fill(
+        stake: &mut StableStake,
+        coin: &mut RegulatedCoin<REST>,
+        value: u64,
+        ctx: &mut TxContext
+    ) {
+        let to_fill = Balance::split(REST::borrow_mut(coin, ctx), value);
+        let coin = Coin::mint<STAKE>(value, &mut stake.treasury_cap, ctx);
```

Comment:
Nit: I found the shadowing of the `coin` parameter confusing here--a different name (`stake_coin`?) might help.

Reply:
Got lost in variables and naming. Nice catch!